File: config/luminophore_shell/monitor_settings.py
"""Typed physical-output rules; parsing never queries connected hardware."""
from dataclasses import dataclass
import math
import re
from .settings_store import StoreError


@dataclass(frozen=True)
class MonitorSettings:
    width: int = 0
    height: int = 0
    refresh: float = 60.0
    scale: float = -1.0
    transform: int = 0
    position: tuple[int, int] | None = None
    disabled: bool = False
    vrr: int | None = None


def decode_monitors(raw):
    def fail():
        raise StoreError('monitors.toml: invalid or unsupported output rule')
    def table(value):
        if type(value) is not dict or any(type(k) is not str for k in value): fail()
        return value
    def point(value):
        if type(value) is not list or len(value) != 2 or any(type(v) is not int or abs(v) > 100000 for v in value): fail()
        return tuple(value)
    table(raw)
    if raw.keys() - {'positions', 'outputs'}: fail()
    positions, outputs = table(raw.get('positions', {})), table(raw.get('outputs', {}))
    names = positions.keys() | outputs.keys()
    if len(names) > 32: fail()
    result = {}
    for name in sorted(names):
        if not re.fullmatch(r'[A-Za-z0-9_-]{1,1024}', name): fail()
        value = table(outputs.get(name, {}))
        if value.keys() - {'mode', 'scale', 'transform', 'position', 'disabled', 'vrr'}: fail()
        mode = value.get('mode', 'preferred')
        width, height, refresh = 0, 0, 60.0
        if type(mode) is not str: fail()
        if mode != 'preferred':
            match = re.fullmatch(r'([1-9][0-9]{0,4})x([1-9][0-9]{0,4})@([1-9][0-9]{0,3}(?:\.[0-9]{1,3})?)', mode)
            if not match: fail()
            width, height, refresh = int(match[1]), int(match[2]), float(match[3])
            if width > 32768 or height > 32768 or refresh > 1000: fail()
        scale = value.get('scale', -1.0)
        if 'scale' in value and (type(scale) not in (int, float) or not math.isfinite(scale) or not .25 <= scale <= 8): fail()
        transform = value.get('transform', 0)
        if type(transform) is not int or not 0 <= transform <= 7: fail()
        disabled = value.get('disabled', False)
        if type(disabled) is not bool: fail()
        vrr = value.get('vrr')
        if 'vrr' in value and (type(vrr) is not int or not 0 <= vrr <= 3): fail()
        if name in positions and 'position' in value: fail()
        position = point(positions[name]) if name in positions else point(value['position']) if 'position' in value else None
        result[name] = MonitorSettings(width, height, refresh, float(scale), transform, position, disabled, vrr)
    return result

File: compositor/src/config/luminophore/MonitorSettings.cpp
#include "MonitorSettings.hpp"
#include <toml++/toml.hpp>
#include <cmath>
#include <regex>
#include <set>
#include <stdexcept>

using namespace Luminophore::Settings;
static void require(bool valid) {
    if (!valid)
        throw std::runtime_error("monitors.toml: invalid or unsupported output rule");
}
static std::pair<int, int> point(const toml::node& node) {
    const auto* a = node.as_array();
    require(a && a->size() == 2);
    require(a->at(0).is_integer() && a->at(1).is_integer());
    auto x = *a->at(0).value<int64_t>(), y = *a->at(1).value<int64_t>();
    require(x >= -100000 && x <= 100000 && y >= -100000 && y <= 100000);
    return {static_cast<int>(x), static_cast<int>(y)};
}
MonitorSettings Luminophore::Settings::decodeMonitorSettings(std::string_view document) {
    const auto raw = toml::parse(document);
    require(raw["schema_version"].is_integer() && raw["schema_version"].value<int64_t>() == 1);
    for (const auto& [key, node] : raw)
        require(key == "schema_version" || ((key == "positions" || key == "outputs") && node.is_table()));
    const toml::table     empty;
    const auto&           positions = raw["positions"].is_table() ? *raw["positions"].as_table() : empty;
    const auto&           outputs   = raw["outputs"].is_table() ? *raw["outputs"].as_table() : empty;
    std::set<std::string> names;
    for (const auto& [key, node] : positions)
        names.emplace(key.str());
    for (const auto& [key, node] : outputs)
        names.emplace(key.str());
    require(names.size() <= 32);
    MonitorSettings  result;
    const std::regex identifier("[A-Za-z0-9_-]{1,1024}");
    const std::regex mode("([1-9][0-9]{0,4})x([1-9][0-9]{0,4})@([1-9][0-9]{0,3}(?:\\.[0-9]{1,3})?)");
    for (const auto& name : names) {
        require(std::regex_match(name, identifier));
        SMonitorSettings rule;
        require(!outputs.contains(name) || outputs[name].is_table());
        const auto& fields = outputs.contains(name) ? *outputs[name].as_table() : empty;
        for (const auto& [key, node] : fields)
            require(key == "mode" || key == "scale" || key == "transform" || key == "position" || key == "disabled" || key == "vrr");
        if (fields.contains("mode")) {
            require(fields["mode"].is_string());
            const auto text = *fields["mode"].value<std::string>();
            if (text != "preferred") {
                std::smatch match;
                require(std::regex_match(text, match, mode));
                rule.width   = std::stoi(match[1]);
                rule.height  = std::stoi(match[2]);
                rule.refresh = std::stod(match[3]);
                require(rule.width <= 32768 && rule.height <= 32768 && rule.refresh <= 1000);
            }
        }
        if (fields.contains("scale")) {
            require(fields["scale"].is_integer() || fields["scale"].is_floating_point());
            rule.scale = *fields["scale"].value<double>();
            require(std::isfinite(rule.scale) && rule.scale >= .25 && rule.scale <= 8);
        }
        if (fields.contains("transform")) {
            require(fields["transform"].is_integer());
            const auto transform = *fields["transform"].value<int64_t>();
            require(transform >= 0 && transform <= 7);
            rule.transform = transform;
        }
        if (fields.contains("disabled")) {
            require(fields["disabled"].is_boolean());
            rule.disabled = *fields["disabled"].value<bool>();
        }
        if (fields.contains("vrr")) {
            require(fields["vrr"].is_integer());
            const auto vrr = *fields["vrr"].value<int64_t>();
            require(vrr >= 0 && vrr <= 3);
            rule.vrr = static_cast<int>(vrr);
        }
        require(!(positions.contains(name) && fields.contains("position")));
        if (positions.contains(name))
            rule.position = point(*positions.get(name));
        else if (fields.contains("position"))
            rule.position = point(*fields.get("position"));
        result.emplace(name, rule);
    }
    return result;
}
